File: backend/src/agents/executor.py
import asyncio
from src.agents.base import BaseAgent, AgentProfile, TaskResult
from src.integrations.kalibr_router import routed_completion
from typing import Optional

class ExecutorAgent(BaseAgent):
    def __init__(self):
        super().__init__(AgentProfile(
            agent_id="executor",
            name="Executor",
            role="executor",
            description="Takes on-chain actions: swaps, LP positions, token operations. Requires human approval for high-value txns.",
        ))
        self.approval_threshold = 0.1  # SOL

    async def execute(self, task: dict) -> TaskResult:
        description = task.get("description", "")
        context = task.get("context", [])

        await self.emit("executor.planning", {
            "message": f"Planning execution: {description[:100]}...",
        })
        await asyncio.sleep(0.5)

        context_str = "\n".join([f"- {c['from_agent']}: {c['data']}" for c in context]) if context else "No prior context."

        prompt = f"""You are an on-chain execution agent on Solana. Plan the execution steps.

Task: {description}
Context: {context_str}

Describe:
1. What on-chain action you would take
2. Estimated cost in SOL
3. Risk level
4. Expected outcome

Note: This is a devnet simulation."""

        plan = await routed_completion(
            goal="execution_plan",
            messages=[{"role": "user", "content": prompt}],
        )

        await self.emit("executor.executing", {
            "message": "Executing on-chain action (devnet)",
            "plan_summary": plan[:200],
        })
        await asyncio.sleep(0.5)

        # Simulate transaction
        await self.emit("executor.tx_submitted", {
            "message": "Transaction submitted to Solana devnet",
            "status": "confirmed",
            "explorer_url": "https://explorer.solana.com/?cluster=devnet",
        })

        return TaskResult(
            success=True,
            data={"execution_plan": plan, "status": "simulated", "network": "devnet"},
            cost=0.005,
        )

    async def bid(self, task: dict) -> Optional[float]:
        if task.get("type") == "execution":
            return 0.005
        return None

File: frontend/src/lib/agents/executor.ts
import { BaseAgent } from "./base";
import { TaskResult, TaskInput } from "./types";
import { llmCompletion } from "../llm";

export class ExecutorAgent extends BaseAgent {
  approvalThreshold = 0.1; // SOL

  constructor() {
    super({
      agentId: "executor",
      name: "Executor",
      role: "executor",
      description: "Takes on-chain actions: swaps, LP positions, token operations. Requires human approval for high-value txns.",
    });
  }

  async execute(task: TaskInput): Promise<TaskResult> {
    const description = task.description || "";
    const context = task.context || [];

    this.emit("executor.planning", {
      message: `Planning execution: ${description.slice(0, 100)}...`,
    });

    const contextStr = context.length
      ? context.map((c) => `- ${c.from_agent}: ${c.data}`).join("\n")
      : "No prior context.";

    const systemPrompt = `You are an on-chain execution agent operating on Solana. Plan and describe the execution steps in detail.

Context from other agents:
${contextStr}

Structure your response as:
## Execution Plan
### Steps (numbered, each with a bold action name and specific details — addresses, amounts, slippage settings)
### Estimated Costs (itemized transaction fees in SOL)
### Risk Level (with slippage protection details)
### Expected Outcome (specific numbers — APY, position size, expected returns)

Note: This is a devnet simulation. Be specific about which programs/protocols you'd interact with (Jupiter, Orca, Raydium, etc.).`;

    const plan = await llmCompletion("execution_plan", [
      { role: "system", content: systemPrompt },
      { role: "user", content: description },
    ]);

    this.emit("executor.executing", {
      message: "Executing on-chain action (devnet)",
      plan_summary: plan.slice(0, 200),
    });

    this.emit("executor.tx_submitted", {
      message: "Transaction submitted to Solana devnet",
      status: "confirmed",
      explorer_url: "https://explorer.solana.com/?cluster=devnet",
    });

    return {
      success: true,
      data: { execution_plan: plan, status: "simulated", network: "devnet" },
      cost: 0.0005,
    };
  }

  async bid(task: TaskInput): Promise<number | null> {
    return task.type === "execution" ? 0.0005 : null;
  }
}
